### PR TITLE
Android cashlib

### PR DESCRIPTION
**Pull request by gandrewstone**

### CODE REVIEW

**Comment by sickpig** [comment]
@gandrewstone why you needed to merge `dev` into your local `andoridCashlib` branch? especially because the only change that actuially matters is this one: 

```diff
  #include "primitives/transaction.h"
  #include "serialize.h"
- // #include "sync.h"
  #include "uint256.h"
  class arith_uint256;
```

File: src/pow.cpp [hunk 1]
```diff
@@ -161,21 +161,7 @@ bool CheckProofOfWork(uint256 hash, unsigned int nBits, const Consensus::Params
     return true;
 }
 
-arith_uint256 GetBlockProof(const CBlockIndex &block)
```
Comment (Greg-Griffith) [inline src/pow.cpp:164]: Are the changes you made WRT this function due to linker issues when compiling cashlib? If not i dont understand the reason for this change

**Comment by Greg-Griffith** [comment]
Can you add a diff file to this PR for the changes you have made to libsecp256k1 in the event we need to look it up later?

**Comment by sickpig** [comment]
@gandrewstone I think that the Andorid test code should be moved under a new separated repo. Basically everything under `cashlib/Test_libbitcoincash_android`

File: src/cashlib/libsecp256k1-config.h [hunk 2]
```diff
@@ -0,0 +1,163 @@
+/* src/libsecp256k1-config.h.  Generated from libsecp256k1-config.h.in by configure.  */
+/* src/libsecp256k1-config.h.in.  Generated from configure.ac by autoheader.  */
```
Comment (sickpig) [inline src/cashlib/libsecp256k1-config.h]: why do we need to have this file if it is generated by the `configure` process?

File: src/secp256k1/src/field_impl.h [hunk 3]
```diff
@@ -13,9 +13,9 @@
 
 #include "util.h"
 
-#if defined(USE_FIELD_10X26)
+#if defined (USE_FIELD_10X26)
 #include "field_10x26_impl.h"
-#elif defined(USE_FIELD_5X52)
+#elif defined (USE_FIELD_5X52)
 #include "field_5x52_impl.h"
```
Comment (sickpig) [inline src/secp256k1/src/field_impl.h]: unrelated change

File: src/secp256k1/src/num.h [hunk 4]
```diff
@@ -7,12 +7,12 @@
...
 #define SECP256K1_NUM_H
 
-#ifndef USE_NUM_NONE
-
 #if defined HAVE_CONFIG_H
 #include "libsecp256k1-config.h"
 #endif
 
+#ifndef USE_NUM_NONE
+
```
Comment (sickpig) [inline src/secp256k1/src/num.h:15]: why is this necessary?

File: src/bandb.cpp [hunk 5]
```diff
@@ -6,7 +6,8 @@
 
 #include "bandb.h"
 #include "chainparams.h"
-#include "hash.h"
+#include "fs.h"
```
Comment (Greg-Griffith) [inline src/bandb.cpp]: why was fs.h added here?

File: src/cashlib/libsecp256k1-config.h (hunk 2 above)
Comment (gandrewstone) [reply src/cashlib/libsecp256k1-config.h]: I'll remove that comment.  The android config.h is not generated by configure.  I don't run configure at all since it seems to be incompatible (or at least a huge waste of time making it so) with the gradle/cmake android build process.

File: src/pow.cpp (hunk 1 above)
Comment (gandrewstone) [reply src/pow.cpp:164]: yes.  I wanted to remove CBlockIndex from the android build, but still provide this functionality.  The only use of CBlockIndex in this function is to get the "nBits" field.  So, without changing semantics, I refactored this call into a new function GetWorkForDifficultyBits that just takes the nBits value.  cashlib pulls in GetWorkForDifficultyBits, but not this function.

  There are other dependencies I didn't want in pow.cpp, so I placed the new function into block.cpp, which is not a bad place for it because its calculating a block's work.

**Comment by gandrewstone** [comment]
@Greg-Griffith EDIT: oops, there are a few small changes (not to logic).  I'll add a diff.

File: src/bandb.cpp (hunk 5 above)
Comment (gandrewstone) [reply src/bandb.cpp]: seems unnecessary, I'll remove

**Comment by Greg-Griffith** [comment]
> @Greg-Griffith there aren't any changes to secp256k1, except for an accidental additional space that I will remove today

yeah, looking again i see they are all simply spacing changes except for one ifdef that was moved (USE_NUM_NONE).

File: src/secp256k1/src/num.h (hunk 4 above)
Comment (gandrewstone) [reply src/secp256k1/src/num.h:15]: because libsecp256k1-config.h might define USE_NUM_NONE

File: src/bandb.cpp (hunk 5 above)
Comment (gandrewstone) [reply src/bandb.cpp]: removed

File: src/secp256k1/src/field_impl.h (hunk 3 above)
Comment (gandrewstone) [reply src/secp256k1/src/field_impl.h]: removed

File: src/secp256k1/src/num.h (hunk 4 above)
Comment (sickpig) [reply src/secp256k1/src/num.h:15]: Then why upstream kept the original order?
https://github.com/bitcoin-core/secp256k1/blob/master/src/num.h#L10

did we change something else in  `libsecp256k1-config.h.in` that requires it?

File: src/cashlib/Test_libbitcoincash_android/app/src/androidTest/java/info/bitcoinunlimited/www/test_libbitcoincash_android/InstrumentedTest.kt [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+val LogIt = Logger.getLogger("AndroidTest")
+
+// Troubleshooting:
+
+// ---
+// native compile: gmp.h not found
+
+// you have a generated file "src/secp256k1/src/libsecp256k1-config.h".  This was generated for your native build.  You need to use the one in src/cashlib.  To do so, just delete
+// "src/secp256k1/src/libsecp256k1-config.h".  Tip: if you do out-of-source native builds you will be able to have both android and normal builds simultaneously.
```
Comment (sickpig) [inline src/cashlib/Test_libbitcoincash_android/app/src/androidTest/java/info/bitcoinunlimited/www/test_libbitcoincash_android/InstrumentedTest.kt]: so it means that we now have 2 version of `libsecp256k1-config.h`, one for normal build and one for android. 
Is it possible to use 2 different file names? 
Isn't this is a good reason to have this code in a separate repo that link back to BUcash code via, e.g., `git subtree`?

File: src/Makefile.am [hunk 7]
```diff
@@ -600,7 +619,8 @@ endif
 
 libbitcoincash_la_LDFLAGS = $(AM_LDFLAGS) $(SSL_LIBS) $(CRYPTO_LIBS) $(BOOST_LIBS) -no-undefined $(RELDFLAGS)
 libbitcoincash_la_LIBADD = $(LIBSECP256K1)
-libbitcoincash_la_CPPFLAGS = $(AM_CPPFLAGS) -I$(BOOST_CPPFLAGS) -I$(builddir)/obj -I$(srcdir)/secp256k1/include -DBUILD_BITCOIN_INTERNAL
+# libbitcoincash_la_CPPFLAGS = $(AM_CPPFLAGS) -I$(BOOST_CPPFLAGS) -I$(builddir)/obj -I$(srcdir)/secp256k1/include -DBUILD_BITCOIN_INTERNAL
+libbitcoincash_la_CPPFLAGS = $(AM_CPPFLAGS) -I$(BOOST_CPPFLAGS) -I$(builddir)/obj -I$(srcdir)/secp256k1/include -I/usr/lib/jvm/java-8-openjdk-amd64/include/linux -I/usr/lib/jvm/java-8-openjdk-amd64/include -DBUILD_BITCOIN_INTERNAL
```
Comment (sickpig) [inline src/Makefile.am]: hard code path. this mean you'll be able to build android stuff only if you have open jdk installed installed on that specific path. It also implies that you can only build it on linux based sys.

File: src/bloom.cpp [hunk 8]
```diff
@@ -182,6 +182,7 @@ bool CBloomFilter::IsWithinSizeConstraints() const
     return vData.size() <= SMALLEST_MAX_BLOOM_FILTER_SIZE && nHashFuncs <= MAX_HASH_FUNCS;
 }
 
+#ifndef ANDROID // We do not want to pull "Solver" into the Android cashlib compile
```
Comment (sickpig) [inline src/bloom.cpp:185]: why not?

File: src/bloom.cpp [hunk 9]
```diff
@@ -366,7 +368,9 @@ bool CRollingBloomFilter::contains(const uint256 &hash) const
 bool CRollingBloomFilter::contains(const COutPoint &outpoint) const { return contains(ToVector(outpoint)); }
 void CRollingBloomFilter::reset()
 {
+#ifndef ANDROID // On Android don't pick a new tweak value because we don't have GetRand
     nTweak = GetRand(std::numeric_limits<unsigned int>::max());
```
Comment (sickpig) [inline src/bloom.cpp:376]: is this because OpenSSL `RAND_bytes` is not available? Isn't there any other method to get a random value while building for Andorid? 
this discussion seems relevant: https://github.com/arvidn/libtorrent/issues/3103

File: src/cashlib/cashlib.cpp [hunk 10]
```diff
@@ -1,26 +1,45 @@
...
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
-
 /* clang-format off */
 // must be first for windows
 #include "compat.h"
-
 /* clang-format on */
+
+#define JAVA
```
Comment (sickpig) [inline src/cashlib/cashlib.cpp]: why we are doing it unconditionally?

File: src/cashlib/cashlib.cpp [hunk 11]
```diff
@@ -1,26 +1,45 @@
...
+#include "cashaddrenc.h"
+#include "chainparams.h"
 #include "random.h"
 #include "streams.h"
 #include "uint256.h"
 #include "util.h"
 #include "utilstrencodings.h"
 
+//#include "secp256k1.h"
+//#include "secp256k1_recovery.h"
```
Comment (sickpig) [inline src/cashlib/cashlib.cpp]: why are the above commented?

File: src/cashlib/cashlib.cpp [hunk 12]
```diff
@@ -76,24 +110,13 @@ SLAPI int Bin2Hex(unsigned char *val, int length, char *result, unsigned int res
     return 1;
 }
 
-
-/** Return random bytes from cryptographically acceptable random sources */
-SLAPI int RandomBytes(unsigned char *buf, int num)
-{
```
Comment (sickpig) [inline src/cashlib/cashlib.cpp:82]: This remove RandomBytes from cashlib for all targets not just android. I don't think it is correct?

File: src/cashlib/cashlib.cpp [hunk 13]
```diff
@@ -76,24 +110,13 @@ SLAPI int Bin2Hex(unsigned char *val, int length, char *result, unsigned int res
...
-    return num;
-}
-
 /** Given a private key, return its corresponding public key */
 SLAPI int GetPubKey(unsigned char *keyData, unsigned char *result, unsigned int resultLen)
 {
     if (!sigInited)
     {
         sigInited = true;
+        SHA256AutoDetect();
```
Comment (sickpig) [inline src/cashlib/cashlib.cpp:115]: Seems unrelated to this PR

File: src/cashlib/cashlib.cpp [hunk 14]
```diff
@@ -477,3 +501,460 @@ SLAPI unsigned int SmGetError(void *smId)
...
+
+
+// result must be 32 bytes
+extern "C" void hash256(const unsigned char *data, unsigned char len, unsigned char *result)
+{
+    CHash256 hash;
+    hash.Write(data, len);
+    hash.Finalize(result);
+}
+
```
Comment (sickpig) [inline src/cashlib/cashlib.cpp:517]: could be wrong but `hash256` and `sha256` seems almost identical to me. Do we need both?

File: src/cashlib/libsecp256k1-config.h [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+
+/* Define if building universal (internal helper macro) */
+/* #undef AC_APPLE_UNIVERSAL_BUILD */
+
+/* Define this symbol to compile out all VERIFY code */
+/* #undef COVERAGE */
+
+/* Define this symbol to enable the ECDH module */
+/* #undef ENABLE_MODULE_ECDH */
+
```
Comment (sickpig) [inline src/cashlib/libsecp256k1-config.h:17]: since it is a new file why keeping around all those commented LOC?

File: src/fastfilter.h [hunk 16]
```diff
@@ -49,8 +49,9 @@ class CVariableFastFilter
         nFilterBytes = (uint32_t)(std::ceil(-1 / LN2SQUARED * nElements * log(nFPRate) / 8));
         nFilterItems = 8 * nFilterBytes;
 
-        if (nFilterBytes > std::numeric_limits<uint32_t>::max())
-            throw std::runtime_error("CVariableFastFilter can have size no greater maximum uint32_t.");
+        // Not possible because nFilterBytes is uint32_t
+        // if (nFilterBytes > std::numeric_limits<uint32_t>::max())
+        //    throw std::runtime_error("CVariableFastFilter can have size no greater maximum uint32_t.");
```
Comment (sickpig) [inline src/fastfilter.h]: I'd rather have it on another PR.

File: src/tweak.h [hunk 17]
```diff
@@ -40,37 +40,37 @@ class CTweakBase
 inline void fill(const UniValue &v, double &output)
 {
     if (v.isStr())
-        output = boost::lexical_cast<double>(v.get_str());
+        output = std::stod(v.get_str());
```
Comment (sickpig) [inline src/tweak.h]: I'd rather have this change and all the changes below on a separate PR.

File: src/util.h [hunk 18]
```diff
@@ -28,8 +28,48 @@
...
+inline std::string _(const char *psz)
+{
+    std::string rv = translationInterface.Translate(psz);
+    if (rv.empty())
+        return std::string(psz);
+    return rv;
+}
+#endif
+
+// #include <boost/thread/exceptions.hpp>
```
Comment (sickpig) [inline src/util.h]: why `exceptions.hpp` has been commented?

File: src/fs.h [hunk 19]
```diff
@@ -21,4 +21,20 @@ namespace fsbridge {
     FILE *freopen(const fs::path& p, const char *mode, FILE *stream);
 };
 
+bool RenameOver(fs::path src, fs::path dest);
```
Comment (sickpig) [inline src/fs.h:25]: sensible change even if I'd rather have those on a separate PR.

File: src/support/pagelocker.cpp [hunk 20]
```diff
@@ -30,7 +30,7 @@
 #endif
 
 LockedPageManager* LockedPageManager::_instance = NULL;
-boost::once_flag LockedPageManager::init_flag = BOOST_ONCE_INIT;
+std::once_flag LockedPageManager::init_flag;
```
Comment (sickpig) [inline src/support/pagelocker.cpp]: I'd rather have those on a separate PR.
don't want to be pedantic. But there's really a lot of good and isolated changes in this PR that should be merged via separate changes, IMHO.

**Comment by sickpig** [comment]
@gandrewstone 

Travis is failing to compile the source code

```
 CXX      cashlib/libbitcoincash_la-cashlib.lo
../../src/cashlib/cashlib.cpp:546:10: fatal error: jni.h: No such file or directory
 #include <jni.h>
          ^~~~~~~
compilation terminated.
```

seems like a missed `if(n)def` somewhere.

**Comment by sickpig** [comment]
@gandrewstone 

I think that in any case (either we merged the code as is or decide to store it in another repo) we lack of the needed documentations on how to get cashlib compiled for android (list of dependencies, tools, etc etc)

**Comment by sickpig** [comment]
@gandrewstone ping

File: src/bloom.cpp (hunk 9 above)
Comment (gandrewstone) [reply src/bloom.cpp:376]: Note: remove the entire rolling bloom filter from java/android since its not being used.

File: src/bloom.cpp (hunk 8 above)
Comment (gandrewstone) [reply src/bloom.cpp:185]: its not used right now -- so leave it out to minimize dependencies

File: src/cashlib/Test_libbitcoincash_android/app/src/androidTest/java/info/bitcoinunlimited/www/test_libbitcoincash_android/InstrumentedTest.kt (hunk 6 above)
Comment (gandrewstone) [reply src/cashlib/Test_libbitcoincash_android/app/src/androidTest/java/info/bitcoinunlimited/www/test_libbitcoincash_android/InstrumentedTest.kt]: It needs to be the same file name because that's the header name included in all the files.  We do have one for a normal build and one for android because android is a crossbuild and this file basically lets you select architecture-specific optimizations.  My android version selects no optimizations.  Someday we could make this better, but we don't really need optimized signature generation or checking on Android right now.

File: src/bloom.cpp (hunk 9 above)
Comment (gandrewstone) [reply src/bloom.cpp:376]: done

File: src/cashlib/cashlib.cpp (hunk 11 above)
Comment (gandrewstone) [reply src/cashlib/cashlib.cpp]: removed

File: src/cashlib/cashlib.cpp (hunk 12 above)
Comment (gandrewstone) [reply src/cashlib/cashlib.cpp:82]: Its just moved lower in the file so I can provide a different Android implementation.

File: src/cashlib/cashlib.cpp (hunk 14 above)
Comment (gandrewstone) [reply src/cashlib/cashlib.cpp:517]: hash256 is a double sha256.  I don't want to get too deeply into what we "need" and into the wisdom of the hash256 name choice which is causing this confusion.  I wanted to preserve the terminology used in bitcoin code and dev documentation and move on.

File: src/cashlib/libsecp256k1-config.h (hunk 15 above)
Comment (gandrewstone) [reply src/cashlib/libsecp256k1-config.h:17]: for documentation.  they describe the options we might choose in the future

File: src/util.h (hunk 18 above)
Comment (gandrewstone) [reply src/util.h]: its not needed, I will remove

File: src/support/pagelocker.cpp (hunk 20 above)
Comment (gandrewstone) [reply src/support/pagelocker.cpp]: submitted #1745

File: src/tweak.h (hunk 17 above)
Comment (gandrewstone) [reply src/tweak.h]: taken out of this PR and already merged in 1.6.0.1

File: src/fastfilter.h (hunk 16 above)
Comment (gandrewstone) [reply src/fastfilter.h]: change removed

File: src/Makefile.am (hunk 7 above)
Comment (gandrewstone) [reply src/Makefile.am]: Fixed.  to enable java bindings for cashlib there is now a flag in configure, and the JAVA_HOME env var is used to discover which java to use.  JAVA_HOME is the standard way to do this in java.

File: src/cashlib/Test_libbitcoincash_android/app/src/androidTest/java/info/bitcoinunlimited/www/test_libbitcoincash_android/InstrumentedTest.kt (hunk 6 above)
Comment (gandrewstone) [reply src/cashlib/Test_libbitcoincash_android/app/src/androidTest/java/info/bitcoinunlimited/www/test_libbitcoincash_android/InstrumentedTest.kt]: The android cashlib test was removed from this PR since it was agreed that its just too much extraneous "stuff" to add to the repo right now.  The functionality will be tested in another repo.

File: src/secp256k1/src/num.h (hunk 4 above)
Comment (gandrewstone) [reply src/secp256k1/src/num.h:15]: I've configured the android version of libsecp256k1 to have no architecture specific optimizations because the android code is not performance sensitive and bringing architecture-specific opts will not be easy since android runs on many architectures.  This no-opts configuration is likely extremely rarely used in libsecp256k1.

File: src/cashlib/cashlib.cpp (hunk 13 above)
Comment (gandrewstone) [reply src/cashlib/cashlib.cpp:115]: This PR also fixes some issues with cashlib that were discovered while using it in Android and Java.

File: src/util.h (hunk 18 above)
Comment (gandrewstone) [reply src/util.h]: done

File: src/cashlib/cashlib.cpp [hunk 21]
```diff
@@ -1,26 +1,41 @@
...
 #include "uint256.h"
 #include "util.h"
 #include "utilstrencodings.h"
 
+#define MAX_SIG_LEN 100 // DER-encoded ECDSA is more like 72 but better to be safe
+
+#ifndef ANDROID
 #include <openssl/rand.h>
+#endif
+
```
Comment (ptschip) [inline src/cashlib/cashlib.cpp:30]: There should be some setup instructions in /doc/build-aux for setting up the build environment for android

ie.  sudo apt-get install default-jre,  sudo apt-get install default-jdk
and setting the JAVA_HOME env var etc...

File: src/rpc/blockchain.cpp [hunk 22]
```diff
@@ -808,6 +808,24 @@ UniValue gettxoutsetinfo(const UniValue &params, bool fHelp)
     return ret;
 }
 
+UniValue evicttransaction(const UniValue &params, bool fHelp)
+{
+    if (fHelp || params.size() < 1)
```
Comment (ptschip) [inline src/rpc/blockchain.cpp:813]: should probably add a simple test for this in one of the mempool_* python scripts.

Comment (Greg-Griffith) [reply src/rpc/blockchain.cpp:813]: Seconded that it would be nice, but i don't consider not having one a reason to block the merge

**Comment by Greg-Griffith** [comment]
this is just a nit / quality of life thing

@gandrewstone in files such as cashlib.cpp where there are a lot of preprocessor defines (ifdef, ifndef, etc.) can you add comments to the else and endifs to mark what section you are ending if the define lasts for more than a few lines. it can get hard to follow if there are a lot of them especially if some are nested.

**Review by ptschip — APPROVED** [review]
